tests/test_brokertypes.py

```diff
@@ -0,0 +1,473 @@
+
+        self.assertNotEqual(Port(10), Port(20))
+        self.assertNotEqual(Port(10), Port(10, Port.Proto.UDP))
+        self.assertNotEqual(Port(10, Port.Proto.UDP), Port(10, Port.Proto.ICMP))
+        self.assertNotEqual(Port(10, Port.Proto.UDP), Port(10, Port.Proto.UNKNOWN))
+
+        with self.assertRaises(ValueError):
+            Port('oops', Port.Proto.TCP)
+        with self.assertRaises(ValueError):
+            Port(70000)
+        with self.assertRaises(TypeError):
+            Port(10, 'tcp')
+
+        self.assertTrue(Port(10) < Port(20))
+        self.assertTrue(Port(20) > Port(10))
+        self.assertTrue(Port(20, Port.Proto.TCP) < Port(10, Port.Proto.UDP))
+        self.assertTrue(Port(20, Port.Proto.UDP) < Port(10, Port.Proto.ICMP))
+        self.assertHash(Port(10))
+
+    def test_vector(self):
+        val = Vector([from_py(1), from_py('foo'), from_py(True)])
+
+        self.assertEqual(val, val)
+        self.assertEqual(Vector([String('foo')]).to_py(), ['foo'])
+        self.assertEqual(Vector([String('foo')]), from_py(['foo']))
+
+        self.assertNotEqual(Vector([from_py(1), from_py('foo')]),
+                            Vector([from_py(1)]))
+        self.assertNotEqual(Vector([from_py(1), from_py('foo')]),
+                            Vector([from_py(1), from_py('noo')]))
+
+        self.assertEqualRoundtrip(Vector([from_py(1), from_py('foo'), from_py(True)]))
+
+        for _ in val:
+            pass
+        self.assertEqual(len(val), 3)
+        self.assertEqual(val[0], Integer(1))
+
+        self.assertTrue(Vector([from_py(1)]) < Vector([from_py(2)]))
+        self.assertTrue(Vector([from_py(1)]) < Vector([from_py(1), from_py('foo')]))
+        self.assertHash(val)
+
+        for val in (23, [23]):
+            with self.assertRaises(TypeError):
+                Vector(val)
+
+    def test_set(self):
+        val = Set({from_py(1), from_py('foo'), from_py(True)})
+
+        self.assertEqual(val, val)
+        self.assertEqual(Set({String('foo')}).to_py(), {'foo'})
+        self.assertEqual(Set({String('foo')}), from_py({'foo'}))
+
+        self.assertNotEqual(Set({from_py(1), from_py('foo')}),
+                            Set({from_py(1)}))
+        self.assertNotEqual(Set({from_py(1), from_py('foo')}),
+                            Set({from_py(1), from_py('noo')}))
+
+        self.assertEqualRoundtrip(Set({from_py(1), from_py('foo'), from_py(True)}))
+
+        for _ in val:
+            pass
+        self.assertEqual(len(val), 3)
+        self.assertTrue(Integer(1) in val)
+
+        self.assertTrue(Set({from_py(1)}) < Set({from_py(2)}))
+        self.assertTrue(Set({from_py(1)}) < Set({from_py(1), from_py('foo')}))
+        self.assertHash(val)
+
+        for val in (23, {23}):
+            with self.assertRaises(TypeError):
+                Set(val)
+
+    def test_table(self):
+        val = Table({from_py('foo'): from_py(1),
+                     from_py('bar'): from_py(2)})
```

Comment:
General question: Could we make those `from_py()` calls implicit if one just passes Python types?

Reply:
Good one, yeah. I went back and forth over this and in the end figured that's a contract for the current brokertypes API: when constructing complex objects (tables, etc) in there, any existing objects need to be brokertypes already. It seemed to simplify some cases where the API otherwise gets too "squishy", with lots of type inference etc.

I was thinking that when/if this becomes its own module, what would type hinting actually look like. That too seems easier this way.